crates/biome_js_analyze/src/lint/style/no_parameter_assign.rs

```diff
@@ -64,48 +116,142 @@ declare_lint_rule! {
                     | AnyJsBindingDeclaration::TsPropertyParameter(_)
             ) {
-                return binding.all_writes(ctx.model());
+                let param_reassignments: Vec<_> = binding
+                    .all_writes(model)
+                    .map(ProblemType::ParameterReassignment)
+                    .collect();
+
+                signals.extend(param_reassignments);
             }
         }
-        // Empty iterator that conforms to `AllBindingWriteReferencesIter` type.
-        std::iter::successors(None, |_| None)
+
+        signals
     }
 
-    fn diagnostic(ctx: &RuleContext<Self>, reference: &Self::State) -> Option<RuleDiagnostic> {
-        let param = ctx.query();
-        Some(
-            RuleDiagnostic::new(
-                rule_category!(),
-                reference.syntax().text_trimmed_range(),
-                markup! {
-                    "Reassigning a "<Emphasis>"function parameter"</Emphasis>" is confusing."
-                },
-            )
-            .detail(
-                param.syntax().text_trimmed_range(),
-                markup! {
-                    "The "<Emphasis>"parameter"</Emphasis>" is declared here:"
-                },
-            )
-            .note(markup! {
-                "Use a local variable instead."
-            }),
-        )
+    fn diagnostic(ctx: &RuleContext<Self>, state: &Self::State) -> Option<RuleDiagnostic> {
+        match state {
+            ProblemType::ParameterReassignment(expression) => Some(
+                RuleDiagnostic::new(
+                    rule_category!(),
+                    expression.syntax().text_trimmed_range(),
+                    markup! {
+                        "Reassigning a "<Emphasis>"function parameter"</Emphasis>" is confusing."
+                    },
+                )
+                    .detail(
+                        ctx.query().syntax().text_trimmed_range(),
+                        markup! {
+                        "The "<Emphasis>"parameter"</Emphasis>" is declared here:"
+                    },
+                    )
+                    .note(markup! {
+                    "Use a local variable instead."
+                }),
+            ),
+
+            ProblemType::PropertyAssignment(expression) => {
+                Some(
+                    RuleDiagnostic::new(
+                        rule_category!(),
+                        expression.syntax().text_trimmed_range(),
+                        markup! {
+                        "Assignment to a property of "<Emphasis>"function parameter"</Emphasis>" is confusing."
+                    })
+                        .note(markup! {"Use a local variable instead."}),
+                )
+            }
+        }
     }
 }
+
+fn extract_statement_from_reference(reference: &Reference) -> Option<JsExpressionStatement> {
+    reference
+        .syntax()
+        .ancestors()
+        .find_map(AnyJsStatement::cast)
```

Comment:
We could use a more restrictive approach to avoid false positives.
Also, this could avoid going until the root in the event where there are no assignment.

```suggestion
        .skip_while(|node| matches!(node.kind, JsSyntaxKind::JS_IDENTIFIER_EXPRESSION | JsSyntaxKind::JS_STATIC_MEMBER_EXPRESSION | JsSyntaxKind::JS_COMPUTED_MEMBER_EXPRESSION))
        .filter_map(AnyJsStatement::cast)
```

Note: I may have forgotten some node types.

Reply:
Also, we should catch more cases as the ones suggested by ESLint:

```js
const f = function(bar) {
    bar.aaa++;
}
const g = function(bar) {
    bar.aaa--;
}
```

To do so, we could declare our own union of nodes and use it instead of `AnyJsAssignment`.
For example:

```rs
declare_node_union! {
    pub AnyJsAssignmentLike =
        JsPostUpdateExpression
        | JsPreUpdateExpression
        | JsAssignmentExpression
}
```